src/project/manifest.rs

```diff
@@ -14,8 +17,554 @@ use serde_with::{serde_as, DeserializeAs, DisplayFromStr, PickFirst};
 use std::collections::HashMap;
 use std::ops::Range;
 use std::path::{Path, PathBuf};
+use toml_edit::{Array, Document, Item, Table, TomlError, Value};
 use url::Url;
 
+/// Handles the project's manifest file.
+/// This struct is responsible for reading, parsing, editing, and saving the manifest.
+/// It encapsulates all logic related to the manifest's TOML format and structure.
+/// The manifest data is represented as a [`ProjectManifest`] struct for easy manipulation.
+/// Owned by the [`Project`] struct, which governs its usage.
+///
+#[derive(Debug, Clone)]
+pub struct Manifest {
+    /// The path to the manifest file
+    pub path: PathBuf,
+
+    /// The raw contents of the manifest file
+    pub contents: String,
+
+    /// Editable toml document
+    pub document: toml_edit::Document,
+
+    /// The parsed manifest
+    pub parsed: ProjectManifest,
+}
+
+impl Manifest {
+    /// Create a new manifest from a path
+    pub fn from_path(path: PathBuf) -> miette::Result<Self> {
```

Comment:
I guess I would just pass it a &Path and make it owned in the function :)

Reply:
But then you might introduce an additional copy right? If you have a `PathBuf` already, you take its reference and then clone it in the function. I think if the function is going to take ownership it makes sense the pass the `PathBuf` straight away. Or use `impl Into<PathBuf>`.

Reply:
Changed it!

Reply:
Scratch my comment, the function doesn't take ownership of the `path`.

Reply:
Oh only now see @baszalmstra comment.

Reply:
Moved it to: 
```rust
pub fn from_path(path: impl AsRef<Path>) -> miette::Result<Self> 
```
As this should avoid unneeded copying and I added a test that shows it is working + also works with multiple types.